### ISSUE TITLE
Twitter Card Missing

### ISSUE THREAD

**Issue by neodjandre2** [sybrew/the-seo-framework#179]
Output Twitter meta tags?
Output various tags targetted at Twitter.

I have checked this option - but I don't see any Twitter Meta Cards being added in the source code. 

Also I am not seeing an image for the post, when sharing via twitter. 

What can be wrong here?

thanks,
Andrew

**Comment by LeBaux** [comment]
Posting the link in question would be beneficial, Andrew 😉

**Comment by sybrew** [comment]
[These two functions](https://github.com/sybrew/the-seo-framework/blob/2.9.3/inc/classes/generate.class.php#L345-L415) determine whether the Twitter Card output is available.

Simplified requirements:
1. At least a social description must be available.
2. At least a social title must be available.
3. An image for `summary_large_image` instead of `summary`.

How-to spec the requirements:
1. This can't be done automatically when you're using a page builder. Fill in a manual description.
2. This can never fail... unless you use improper filters.
3. An image is tried to be taken from many places. However, you can set up a social fallback image to guarantee output.

In short:
1. Fill in a manual description 😄

**Comment by neodjandre2** [comment]
This is an example of problematic link: http://bit.ly/2ux1h5J

The OG properties are output correctly but as you can see there are no tags for twitter at all.

**Comment by neodjandre2** [comment]
If I fill in, manual description it works! 

Is there a function I could use to automatically do this for my posts ?

**Comment by sybrew** [comment]
Awesome! 😄 

> Is there a function I could use to automatically do this for my posts?

The plugin's slug is "autodescription", but it does have its limits. If you're using a page builder, then I'm afraid this isn't possible.
In any case, it's always best to forge manual descriptions. So, making a habit out of it isn't all too bad!

**Comment by neodjandre2** [comment]
I actually want this for my posts only and i have an ACF field called "fl_description". So it should be simple enough to write a function to truncate the "fl_description" and pass this on to the SEO custom description ?

**Comment by neodjandre2** [comment]
actually, from the past, someone gave me this code which does exactly that!

`add_filter( 'the_seo_framework_description_output', 'my_description_excerpt', 10, 2 );
add_filter( 'the_seo_framework_ogdescription_output', 'my_description_excerpt', 10, 2 );
add_filter( 'the_seo_framework_twitterdescription_output', 'my_description_excerpt', 10, 2 );

function my_description_excerpt( $description = '', $id = 0 ) {

	$custom_desc = get_post_meta( $id, 'fl_description', true );
  $custom_desc = wp_strip_all_tags( $custom_desc, true );

	if ( $custom_desc ) {

		$custom_desc = the_seo_framework()->trim_excerpt( $custom_desc, mb_strlen( $custom_desc ), 155 );

		if ( method_exists( the_seo_framework_class(), 's_excerpt' ) ) {
			$custom_desc = the_seo_framework()->s_excerpt( $custom_desc );
		}

		$description = the_seo_framework()->escape_description( $custom_desc );
	}

	return $description;
}`

However, it's only when I manually fill the description box for the post that I get the twitter cards.. strange..

**Comment by sybrew** [comment]
Hi @neodjandre2 

I believe this will do, you can add it to your theme's `functions.php` file:
```
add_filter( 'the_seo_framework_custom_field_description', function( $description, $args ) {

	// Return manual TSF description if set.
	if ( $description )
		return $description;

	/**
	 * Try ACF and set if exists.
	 * @link https://www.advancedcustomfields.com/resources/get_field/
	 */
	if ( function_exists( 'get_field' ) ) {
		// Pass ID for SEO bar compatibility.
		$description = get_field( 'fl_description', $args['id'] ) ?: $description;
	}

	return $description;
}, 10, 2 );
```

**Comment by sybrew** [comment]
@neodjandre2 The filters you've used only apply to the front-end.
The example I just gave works from TSF 2.9.0 😄, which covers the whole spectrum.

**Comment by neodjandre2** [comment]
ok in this case. I will replace with your new function but is it possible to  trim_excerpt to 155 characters with the new function please.. I am not good with modifying functions!

**Comment by sybrew** [comment]
@neodjandre2 here goes...

```
add_filter( 'the_seo_framework_custom_field_description', function( $description, $args ) {

	// Return manual TSF description if set.
	if ( $description )
		return $description;

	/**
	 * Try ACF, trim and set if exists.
	 * @link https://www.advancedcustomfields.com/resources/get_field/
	 */
	if ( function_exists( 'get_field' ) ) {
		// Pass ID for SEO bar compatibility.
		$excerpt = get_field( 'fl_description', $args['id'] ) ?: '';

		// Only trim if there's actually something found.
		if ( $excerpt ) {
			$description = the_seo_framework()->trim_excerpt( $excerpt, mb_strlen( $excerpt ), 155 ); 
		}
	}

	return $description;
}, 10, 2 );
```

**Comment by neodjandre2** [comment]
Ok perfect this works fine but at the beginning of my descriptions I always get these characters:

`&lt;p&gt;`

why is that? :)

**Comment by neodjandre2** [comment]
I have added:

`$excerpt = strip_tags(html_entity_decode($excerpt));`

and it seemed to fix that.. :) thanks for all the help !

**Comment by neodjandre2** [comment]
yes your solution is better 👍
